pilot/pkg/networking/core/v1alpha3/listener_test.go

```diff
@@ -40,47 +42,66 @@ func TestOutboundListenerConflict_HTTP(t *testing.T) {
-		buildService(model.ProtocolHTTP, tnow.Add(2*time.Second)))
+		buildService("test1,com", model.ProtocolHTTP, tnow.Add(1*time.Second)),
+		buildService("test2,com", model.ProtocolTCP, tnow),
+		buildService("test3,com", model.ProtocolHTTP, tnow.Add(2*time.Second)))
 }
 
 func TestOutboundListenerConflict_TCP(t *testing.T) {
 	// The oldest service port is HTTP.  We should encounter conflicts when attempting to add the TCP ports. Purposely
 	// storing the services out of time order to test that it's being sorted properly.
 	testOutboundListenerConflict(t,
-		buildService(model.ProtocolTCP, tnow.Add(1*time.Second)),
-		buildService(model.ProtocolHTTP, tnow),
-		buildService(model.ProtocolTCP, tnow.Add(2*time.Second)))
+		buildService("test1.com", model.ProtocolTCP, tnow.Add(1*time.Second)),
+		buildService("test2.com", model.ProtocolHTTP, tnow),
+		buildService("test3.com", model.ProtocolTCP, tnow.Add(2*time.Second)))
 }
 
 func TestOutboundListenerConflict_Unordered(t *testing.T) {
 	// Ensure that the order is preserved when all the times match. The first service in the list wins.
 	testOutboundListenerConflict(t,
-		buildService(model.ProtocolHTTP, tzero),
-		buildService(model.ProtocolTCP, tzero),
-		buildService(model.ProtocolTCP, tzero))
+		buildService("test1.com", model.ProtocolHTTP, tzero),
+		buildService("test2.com", model.ProtocolTCP, tzero),
+		buildService("test3.com", model.ProtocolTCP, tzero))
 }
 
-func testOutboundListenerConflict(t *testing.T, services ...*model.Service) {
-	t.Helper()
-
+func TestMergeOutboundTCPFilterChains(t *testing.T) {
+	services := []*model.Service{
+		buildService("test1.com", model.ProtocolTCP, tnow.Add(1*time.Second)),
+		buildService("test2.com", model.ProtocolTCP, tnow),
+		buildService("test3.com", model.ProtocolTCP, tnow.Add(2*time.Second)),
+	}
 	p := &fakePlugin{}
-	configgen := NewConfigGenerator([]plugin.Plugin{p})
+	listeners := buildOutboundListeners(p, services...)
+	if len(listeners) != 1 {
+		t.Fatalf("expected %d listeners, found %d", 1, len(listeners))
+	}
+	verifyOutboundTCPListenerHostname(t, listeners[0], "test2.com")
 
-	env := buildListenerEnv()
+	oldestService := getOldestService(services...)
+	oldestProtocol := oldestService.Ports[0].Protocol
+	if oldestProtocol != model.ProtocolHTTP && isHTTPListener(listeners[0]) {
+		t.Fatal("expected TCP listener, found HTTP")
+	} else if oldestProtocol == model.ProtocolHTTP && !isHTTPListener(listeners[0]) {
+		t.Fatal("expected HTTP listener, found TCP")
+	}
 
-	var oldestService *model.Service
-	instances := make([]*model.ServiceInstance, len(services))
-	for i, s := range services {
-		instances[i] = &model.ServiceInstance{
-			Service: s,
-		}
-		if oldestService == nil || s.CreationTime.Before(oldestService.CreationTime) {
-			oldestService = s
-		}
+	// The plugin should show 3 that all 3 listeners were created. Their filter chains should be merged, however.
+	if len(p.outboundListenerParams) != 3 {
```

Comment:
Substitute with checking len(listeners[0].FilterChains)  after the entire buildSidecarOutboundListeners thing..

Reply:
done.